### PR TITLE
add alternate init support + AMD GPU support (OpenRC, runit, s6)

### PR DESCRIPTION
**Pull request by Hyper-KVM**
I think I got it all working together now, but dbus checks seem to be an issue with elogind since elogind does not support Get.Unit method, so I'm not 100% sure how to fix that. So I had to bypass the dbus check method for optimus-manager due to this.
The bash check works just fine and runit, systemd, and OpenRC service check methods all work, maybe you could clean the code up a bit if you like.

And yeah, I broke the other pull request by using --tree-filter, oh well. I'm new to this lol

A thank you out to @yaliv and @Kr1ss-XD btw for most of the base work on AMD iGPU support :)

### CODE REVIEW

**Comment by Askannz** [comment]
What distributions have you tested it with ?

I'll be honest: while I appreciate the effort, I'm really on the fence about adding this to optimus-manager. Keeping it compatible with the myriad of desktop environments and login managers out there has already been a hassle, and multiplying this by 3 init systems would make it even worse. There is only so much time I can dedicate to testing before pushing a new release.

In addition, we don't even support popular distributions like Fedora or Debian, even though they could use such a tool and have the same init system as Arch. So catering toe ven more niche distributions feels weird.

**Comment by Hyper-KVM** [comment]
> What distributions have you tested it with ?

So far Artix and Void. 

> I'll be honest: while I appreciate the effort, I'm really on the fence about adding this to optimus-manager. Keeping it compatible with the myriad of desktop environments and login managers out there has already been a hassle, and multiplying this by 3 init systems would make it even worse. There is only so much time I can dedicate to testing before pushing a new release.

Oh yeah I totally understand, I'm willing to be the maintainer of the OpenRC and runit side. whether you want to merge and update version or keep them seperate is fine. But if you do decide to not merge I should probably differentiate the fork a bit to not cause confusion right? 

> In addition, we don't even support popular distributions like Fedora or Debian, even though they could use such a tool and have the same init system as Arch. So catering to even more niche distributions feels weird.

Yeah, Artix could be considered niche and the majority of users do use systemd so I can see where you're coming from. This was more of a personal quest lol since there was nothing like optimus-manager for Artix/Void, I honestly wasn't expecting much but then just kept going and well now its where it is now with full support for systemd even (I did remove systemd stuff before lol). Well anyway, thank you for developing this awesome tool it really makes my experience a lot better :)

**Comment by Mallchad** [comment]
Works for me man when using runit.
I think, I may have hit a hitch when using it with `pci_remove`  as an option but I believe that's a known
issue anyway.
I should note that I'm not sure this could be fully merged since the AUR version would require a fair few extra steps to setup.
Maybe there come be some init auto-detection, or a simple cli setup wizard to help people install it easily.
There is one issue, where it doesn't play nicely with optimus-manager-qt, but I don't think that's necessarily the fault of this project, I can live with it either way.

**Comment by Hyper-KVM** [comment]
> Works for me man when using runit.
> I think, I may have hit a hitch when using it with `pci_remove` as an option but I believe that's a known
> issue anyway.
> I should note that I'm not sure this could be fully merged since the AUR version would require a fair few extra steps to setup.
> Maybe there come be some init auto-detection, or a simple cli setup wizard to help people install it easily.
> There is one issue, where it doesn't play nicely with optimus-manager-qt, but I don't think that's necessarily the fault of this project, I can live with it either way.

I have added init detection into the PKGBUILD plus a display manager selection for runit, please let me know if something isn't working thanks!

File: package/PKGBUILD [hunk 1]
```diff
@@ -42,31 +41,59 @@ build() {
...
+		install -Dm755 optimus-manager/runit/displaymanager/lightdm/run "$pkgdir/etc/runit/sv/lightdm/run"
+		install -Dm755 optimus-manager/runit/displaymanager/optimus-manager/run "$pkgdir/etc/sv/optimus-manager/run"
+	else
+		echo "No Display Manager selected, installing standalone script"
+		install -Dm755 optimus-manager/runit/optimus-manager/run "$pkgdir/etc/runit/sv/optimus-manager/run"
+	fi
+  # openrc
+  elif [ -n $(command -v rc-service) ]; then
+	install="openrc.install"
+	install -Dm755 optimus-manager/openrc/optimus-manager "$pkgdir/etc/init.d/optimus-manager"
```
Comment (Mallchad) [inline package/PKGBUILD:68]: install -Dm755 optimus-manager/openrc/optimus-manager "$pkgdir/etc/init.d/optimus-manager"

File: optimus-manager.conf [hunk 2]
```diff
@@ -34,20 +34,20 @@ pci_reset=no
...
 # "auto" is a special mode that auto-detects if the computer is running on battery
 # and selects a proper GPU mode. See the other options below.
 startup_mode=intel
 # GPU mode to select when startup_mode=auto and the computer is running on battery.
-# Possible values: nvidia, intel, hybrid
+# Possible values: nvidia, intel, amd, hybrid-intel, hybrid-amd
 startup_auto_battery_mode=intel
 # GPU mode to select when startup_mode=auto and the computer is running on external power.
-# Possible values: nvidia, intel, hybrid
+# Possible values: nvidia, intel, amd, hybrid-intel, hybrid-amd
```
Comment (Mallchad) [inline optimus-manager.conf:50]:
```suggestion
# GPU mode to use at computer startup. Possible values: nvidia, intel, amd, hybrid-nvidial, hybrid-amd, auto
# "auto" is a special mode that auto-detects if the computer is running on battery
# and selects a proper GPU mode. See the other options below.
startup_mode=intel
# GPU mode to select when startup_mode=auto and the computer is running on battery.
# Possible values: nvidia, intel, amd, hybrid-intel, hybrid-amd
startup_auto_battery_mode=intel
# GPU mode to select when startup_mode=auto and the computer is running on external power.
# Possible values: nvidia, intel, amd, hybrid-nvidia, hybrid-amd
```
I would assume the hybrid mode is specifying the GPU not the CPU.
Either way `hybrid-intel` assumes that optimus-manager knows which GPU to use, but I don't think this is the intent.

File: package/PKGBUILD [hunk 3]
```diff
@@ -42,31 +41,59 @@ build() {
...
 package() {
-
-  install="optimus-manager.install"
  
   cd "${srcdir}/optimus-manager/"
- 
+  # runit
+  if [ -n $(command -v sv) ]; then
+	install="runit.install"
+	install -Dm755 optimus-manager/runit/optimus-manager/finish "$pkgdir/etc/runit/sv/optimus-manager/finish"
```
Comment (Mallchad) [inline package/PKGBUILD:49]:
```suggestion
	install -Dm755 runit/optimus-manager/finish "$pkgdir/etc/runit/sv/optimus-manager/finish"
```

File: package/PKGBUILD (hunk 1 above)
Comment (Mallchad) [inline package/PKGBUILD:68]:
```suggestion
		install -Dm755 runit/displaymanager/sddm/run "$pkgdir/etc/runit/sv/sddm/run"
		install -Dm755 runit/displaymanager/optimus-manager/run "$pkgdir/etc/sv/optimus-manager/run"
	elif [ "$CONDM" == "2" ]; then
		echo "Installing LightDM runit script"
		install -Dm755 runit/displaymanager/lightdm/run "$pkgdir/etc/runit/sv/lightdm/run"
		install -Dm755 runit/displaymanager/optimus-manager/run "$pkgdir/etc/sv/optimus-manager/run"
	else
		echo "No Display Manager selected, installing standalone script"
		install -Dm755 runit/optimus-manager/run "$pkgdir/etc/runit/sv/optimus-manager/run"
	fi
  # openrc
  elif [ -n $(command -v rc-service) ]; then
	install="openrc.install"
	install -Dm755 openrc/optimus-manager "$pkgdir/etc/init.d/optimus-manager"
```

File: init/s6/PKGBUILD [hunk 4]
```diff
@@ -0,0 +1,25 @@
+pkgname=optimus-manager-s6
```
[the remark this answers is not in the document]
Comment (Hyper-KVM) [reply init/s6/PKGBUILD:1]: I don't even know if this works, I haven't tested s6.

File: optimus-manager.conf (hunk 2 above)
Comment (Hyper-KVM) [reply optimus-manager.conf:50]: I'm quite sure what you mean here?

File: package/PKGBUILD (hunk 1 above)
Comment (Hyper-KVM) [reply package/PKGBUILD]: Applied patch, resolving :)

File: README.md [hunk 5]
```diff
@@ -47,7 +47,25 @@ Naturally, you must have the proprietary nvidia driver installed on your system.
 
 You can install optimus-manager from this AUR package : [optimus-manager](https://aur.archlinux.org/packages/optimus-manager/)
 
-After installation, the optimus-manager daemon should have been started automatically, but you can check its status with `systemctl status optimus-manager.service`.
+For Artix OpenRC or runit you will need to install the corresponding init script package in the init folder using `makepkg -si`
```
[the remark this answers is not in the document]
Comment (Hyper-KVM) [reply README.md]: Updated.

**Comment by Hyper-KVM** [comment]
> All good, I need to verify if the install script will be spoofed by mixed init/logging systems.
> Some corner cases with people's setups, ie runit-systemd might cause problems with installing the wrong script but it needs verification.

Hmm yeah you're right, this could create problems... I have an idea though. Thanks for pointing this out.

File: optimus-manager.conf (hunk 2 above)
Comment (Mallchad) [reply optimus-manager.conf:50]: What I mean is the hybrid modes are a bit vague. `hybrid-amd` seems like it's for if you have 2 dGPU's, one AMD one Nvidia and you use the AMD card as primary, but what about the other way round? and what if you want primary intel but offload to amd? 
It also could cause confusion with intel-amd systems but that's lesser problem.

Comment (Hyper-KVM) [reply optimus-manager.conf:50]: Oh I see, so in the README and general Description of the Project it does state that this is for Optimus laptops and not Intel/AMD hybrid graphics. Which is Intel/NVIDIA or AMD/NVIDIA. actually now that you say that, i need to do a look over the readme as it doesn't specifically state AMD/NVIDIA just intel/NVIDIA, lol. thanks for mentioning this. :)

**Comment by Hyper-KVM** [comment]
So I have everything in seperate branches now, I'm gonna close this one.

**Comment by DakshMiglani** [comment]
@Hyper-KVM how do I use it on runit then?

**Pull request closed without merging**